pkg/controllers/node/termination/controller.go

```diff
@@ -104,12 +120,24 @@ func (c *Controller) finalize(ctx context.Context, node *v1.Node) (reconcile.Res
 		}
 		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
 	}
+	// In order for stateful pods to smoothly migrate from the terminating Node, we wait for VolumeAttachments
+	// to be cleaned up before terminating the node and removing it from the cluster.
+	if err := c.waitForVolumeDetachments(ctx, node, volumeDetachmentsBackoff); err != nil {
+		c.recorder.Publish(terminatorevents.NodeVolumeAttachmentsRemaining(node, err))
+	}
 	// Be careful when removing this delete call in the Node termination flow
 	// This delete call is needed so that we ensure that we don't remove the node from the cluster
 	// until the full instance shutdown has taken place
 	if err := c.cloudProvider.Delete(ctx, nodeclaimutil.NewFromNode(node)); cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
 		return reconcile.Result{}, fmt.Errorf("terminating cloudprovider instance, %w", err)
 	}
+	// This taint confirms node termination so leftover VolumeAttachments are marked as unmounted and can be detached
+	if err := c.terminator.TaintOutOfService(ctx, node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("tainting node as out-of-service, %w", err)
+	}
+	// Wait to make sure AttachDetach controller notices node is out-of-service.
+	_ = c.waitForVolumeDetachments(ctx, node, outOfServiceTaintBackoff)
```

Comment:
FYI on a small cluster it usually takes 3-8 seconds between applying out-of-service taint and deleting the node finalizer for the AttachDetach controller to see that the node had the taint before node is deleted. 

We can either wait on volumeDetachments with shorter backoff here or just wait until all volumeattachments have DeletionTimestamp if we're in a rush to remove node finalizer. 

Today the node object has to stay up long enough for attachdetach controller to see that the node has this out-of-service taint so that it can apply deletion timestamp to volumeattachment.

Reply:
Are we convinced we should be using the `out-of-service` taint here? I'm not aware of what all it does, other than detaching volumes and taint driven pod eviction, but everywhere it's referenced it's emphasized to only use this on nodes which have already been shut off. Do we know why that is? Is it because of potential data corruption caused by detaching volumes from running processes or something else? 

I'm also wondering if we want to be detaching volumes if pods are still running and using them (e.g. if they tolerated the disruption taint). It seems like there's still an opportunity for them to be cleaned up during graceful shutdown.